### PR TITLE
Complete HE depth feature bit implementation

### PR DESCRIPTION
**Pull request by christopheralanwest**
This PR implements code changes needed to implement depth feature bits [1] into HE trigger primitives. The changes are as follows:

- Generation of HE depth feature bit LUTs for use online (5a23ad7842470b9edf20e3081115d448824d7b25, 5a9be22164f0a1f457aa18b1c60811c539cd739c, c42032b8e661afa41279a1d7b31666c1aae29d5d)
- Do not apply peak finding to feature bits, to match the firmware behavior (as a primary use of the depth feature bit algorithm is MIP detection, peak detection would be very noisy); Fix of indexing issues to reflect that depth indexing starts at 1, not 0 (8ff026a6f96a570a7079d198b5e00e4de5730f0e)
- Addition of new algorithms used for technical tests of the generation of the feature bits (9d50b6d7d19acd67600e62643766a5a4fe7383ee)
- Update to LUT generation configuration and a technical change to the formatting of some of the XML tags in the LUTs (68f7bcc5b86a986b05a32b6b70f7492c34a91191, 8b13429049e2b144883ea96e941c78bc3ca68651) to reflect what is currently in use. This is not logically directly related to other commits but is needed more generally for LUT generation. These changes are already in use.

The feature bit code has been tested at 904 using uHTRs that read out only HE [2] and gives perfect agreement between data and emulation. As mentioned at [2], the online software does not yet support the loading of HE feature bit LUTs in uHTRs that read out both HB and HE and so data/emulation comparisons have not been performed for the highest and lowest ieta ranges of HE. I have verified that the LUTs are nevertheless generated for these towers, even though they are not currently used. Since all towers in HE have identical LUTs, no further changes will be needed for the mixed HB/HE readout.

The only changes in PR comparison tests will appear in the feature bits of the HE trigger primitives, in samples that contain an upgraded HE. As these feature bits are not yet used by any downstream quantity, no physical quantity will change.

[1] for details, see pages 23-25 of https://cms-docdb.cern.ch/cgi-bin/DocDB/ShowDocument?docid=12306
[2] http://cmsonline.cern.ch/cms-elog/1063064

### CODE REVIEW

**Comment by thomreis** [comment]
please test

**Comment by christopheralanwest** [comment]
The message that opened this PR neglected to mention an important consequence of the indexing bug that is fixed in 8ff026a6f96a570a7079d198b5e00e4de5730f0e that is needed for understanding the differences in the PR comparison tests. Because the index of 0 corresponds to depth=1, it was previously impossible to fire any emulated feature bits that required contributions in depth=1. Only bit=2 could fire (https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/HcalTrigPrimAlgos/src/HcalFinegrainBit.cc#L15-L23) as the other non-zero bits required a depth=1 contribution explicitly. 

I've listed below all of the differences observed in the PR comparison tests, and how they are a result of the bug fix.

136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM

Hcal/TPTask/FGMsm has more entries because these bits were previously not commissioned in data, and therefore zero. The additional non-zero bits imply additional mismatches.
Hcal/TPTask/FGMsmFraction is a direct consequence of the differences in the FG mismatch plot

136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017

Similar to the behavior in 136.85 but only HEP17 is affected

150.0_HydjetQ_B12_5020GeV_2018+HydjetQ_B12_5020GeV_2018+DIGIHI2018+RECOHI2018+HARVESTHI2018

Previously feature bits 0 and 1 could not fire in HE, generating differences in:

L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccRecdFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb2

158.0_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+HARVESTHI2018PPRECO

L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccRecdFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb2

Similar reasoning as in RelVal 158.0

10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017

L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccRecdFb

Similar reasoning as in RelVal 158.0, but only in HEP17

10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018+NanoFull_2018

L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccRecdFb
L1T/L1TStage2CaloLayer1/HCalDetail/hcalOccSentFb2

Similar reasoning as in RelVal 158.0

**Comment by abrinke1** [comment]
If @rekovic , @thomreis , @nsmith- , @bundocka, and @davignon have no objections, I think this PR is fine from the L1T side.

File: CaloOnlineTools/HcalOnlineDb/src/HcalLutManager.cc [hunk 1]
```diff
@@ -854,6 +854,84 @@ std::map<int, boost::shared_ptr<LutXml> > HcalLutManager::getLinearizationLutXml
   return _xml;
 }
 
+std::map<int, boost::shared_ptr<LutXml> > HcalLutManager::getHEFineGrainLUTs(std::string _tag, bool split_by_crate )
```
Comment (thomreis) [inline CaloOnlineTools/HcalOnlineDb/src/HcalLutManager.cc]: Can the use of boost::shared_ptr be changed to std::shared_ptr or even std::unique_ptr throughout the code?

**Comment by thomreis** [comment]
please test

File: CaloOnlineTools/HcalOnlineDb/src/HcalLutManager.cc (hunk 1 above)
Comment (christopheralanwest) [reply CaloOnlineTools/HcalOnlineDb/src/HcalLutManager.cc]: The migration has been performed for `HcalLutManager.cc/.hh`. The other usage of `boost::shared_ptr` in `CaloOnlineTools/HcalOnlineDb/interface/LMap.h` is referenced in many places elsewhere in CMSSW and so should be resolved in a separate PR.

**Comment by thomreis** [comment]
please test

**Comment by thomreis** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by lpernie** [comment]
+1

**Comment by christopheralanwest** [comment]
The merging of #24847 generated a merge conflict because both PRs modified the same file, but this PR was missing an import statement here: https://github.com/cms-sw/cmssw/pull/24556/files#diff-bef27a33d095d007df922a4fc8a5096aR3

**Comment by thomreis** [comment]
please test

**Comment by christopheralanwest** [comment]
The failed `runtestPhysicsToolsPatAlgos` test is not a direct consequence of this PR, as the [same failure](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc6_amd64_gcc700/CMSSW_10_4_X_2018-10-16-2300/unitTestLogs/PhysicsTools/PatAlgos#/) is seen in the `CMSSW_10_4_X_2018-10-16-2300` IB.

I think that the PatAlgos unit test was only performed as a consequence of merges with commits that are now in the IB. Running `git cms-merge-topic -u 24556` locally on top of `CMSSW_10_4_X_2018-10-16-2300` does not bring in any additional commits so I think that simply re-running the tests will resolve the issue.

**Comment by thomreis** [comment]
please test

**Comment by christopheralanwest** [comment]
The pattern of changes in the comparison tests is the same as in my [summary](https://github.com/cms-sw/cmssw/pull/24556#issuecomment-422106970) from September 17. No commits that could introduce any change in the behavior of this PR have been introduced since that time.

**Comment by christopheralanwest** [comment]
@rekovic , @thomreis , @nsmith- Could you take a look at this PR? Thanks.

**Comment by thomreis** [comment]
+1

**Comment by fabiocos** [comment]
@lpernie @tocheng @ggovi could you please scrutinize this PR?

**Comment by ggovi** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

AlCa had already signed, the latest update does not look to affect their part

**Pull request merged**